source/SquadLeader/app/src/main/java/com/esri/squadleader/controller/AdvancedSymbolController.java

```diff
@@ -76,22 +75,49 @@
      * @param spotReportIcon the Drawable for putting spot reports on the map.
      * @param messageController a MessageController, for sending updates when messages are to be removed,
      *        e.g. in clearLayer or clearAllMessages.
-     * @throws FileNotFoundException if the advanced symbology database is absent or corrupt.
+     * @throws IOException if the advanced symbology database is absent or corrupt or cannot be written
+     *                     to the device.
      */
     public AdvancedSymbolController(
             MapController mapController,
             AssetManager assetManager,
             String symbolDictionaryDirname,
             Drawable spotReportIcon,
-            MessageController messageController) throws FileNotFoundException {
+            MessageController messageController) throws IOException {
         super(mapController);
         this.mapController = mapController;
+        symDictDir = copySymbolDictionaryToDisk(assetManager, symbolDictionaryDirname);
+
+        spotReportLayer = new GraphicsLayer();
+        spotReportLayer.setName(SPOT_REPORT_LAYER_NAME);
+        mapController.addLayer(spotReportLayer);
+        
+        groupLayer = new MessageGroupLayer(SymbolDictionary.DictionaryType.MIL2525C, symDictDir.getAbsolutePath());
+        mapController.addLayer(groupLayer);
+        
+        spotReportSymbol = new PictureMarkerSymbol(spotReportIcon);
+        
+        this.messageController = messageController;
+    }
+
+    /**
+     * Copies the MIL-STD-2525C symbol dictionary from assets to the device's downloads directory if
+     * it is not already there.
+     * @param assetManager the application's AssetManager, from which the advanced symbology database
```

Comment:
I'm not sure I completely understand how the mil2525c.dat and messagetype folder get on the device and/or sdcard  - but since this was already working with geomessages, we can assume it works fine